### PR TITLE
Fix EZP-23733: The delete button in section view/list should be disabled if the section can not be removed

### PR DESCRIPTION
**Pull request by glye**
> Fixes https://jira.ez.no/browse/EZP-23733
> Also requires https://github.com/ezsystems/PlatformUIBundle/pull/415
> Status: Ready for review

Adds method for counting policies using section in limitations.

- [x] Tests

### CODE REVIEW

File: eZ/Publish/Core/Repository/SectionService.php [hunk 1]
```diff
@@ -307,6 +319,10 @@ public function deleteSection(Section $section)
             throw new BadStateException('section', 'section is still assigned to content');
         }
 
+        if ($this->countPoliciesUsing($loadedSection) > 0) {
+            throw new BadStateException('section', 'section is still being used in policy limitations');
```
Comment (pspanja) [inline eZ/Publish/Core/Repository/SectionService.php]: This case should be added to the docblock.

File: eZ/Publish/Core/Persistence/Legacy/Content/Section/Gateway/DoctrineDatabase.php [hunk 2]
```diff
@@ -207,6 +207,50 @@ public function countContentObjectsInSection($id)
...
+        $query->select(
+            $query->alias(
+                $query->expr->count(
+                    $this->dbHandler->quoteColumn('id', 'ezpolicy_limitation')
+                ),
+                'policy_count'
+            )
+        )->from(
+            $this->dbHandler->quoteTable('ezpolicy_limitation')
+        )->leftJoin(
```
Comment (pspanja) [inline eZ/Publish/Core/Persistence/Legacy/Content/Section/Gateway/DoctrineDatabase.php]: AFAICS there is no need for LEFT JOIN.

File: eZ/Publish/Core/Persistence/Legacy/Content/Section/Gateway/DoctrineDatabase.php [hunk 3]
```diff
@@ -207,6 +207,50 @@ public function countContentObjectsInSection($id)
...
+     *
+     * @param int $id
+     *
+     * @return int
+     */
+    public function countPoliciesUsingSection($id)
+    {
+        $query = $this->dbHandler->createSelectQuery();
+        $query->select(
+            $query->alias(
```
Comment (pspanja) [inline eZ/Publish/Core/Persistence/Legacy/Content/Section/Gateway/DoctrineDatabase.php]: Nitpick: no need to aliasing as you use `fetchColumn()` below.

File: eZ/Publish/API/Repository/Tests/SectionServiceTest.php [hunk 4]
```diff
@@ -505,6 +505,35 @@ public function testCountAssignedContents()
...
+        $sectionService = $repository->getSectionService();
+
+        $standardSectionId = $this->generateId('section', 1);
+        /* BEGIN: Use Case */
+        // $standardSectionId contains the ID of the "Standard" section in a eZ
+        // Publish demo installation.
+
+        $standardSection = $sectionService->loadSection($standardSectionId);
+
+        $numberOfAssignedContent = $sectionService->countPoliciesUsing(
```
Comment (pspanja) [inline eZ/Publish/API/Repository/Tests/SectionServiceTest.php]: Misleading variable name, use `$numberOfPolicies` or similar.

File: eZ/Publish/Core/Repository/SectionService.php (hunk 1 above)
Comment (glye) [inline eZ/Publish/Core/Repository/SectionService.php]: Added.

**Comment by glye** [comment]
@pspanja Updated according to feedback, thanks.

**Comment by pspanja** [comment]
TBH I'm wondering if this should be added to the API. For one, in SqlNg the whole policy/limitation DB data structure was [simplified](https://github.com/ezsystems/ezpublish-kernel/blob/sql-ng/eZ/Publish/Core/Persistence/SqlNg/User/Role/Gateway/EzcDatabase.php#L464) in a way that makes query like this (counting policies using the Section) impossible. So accepting this would probably mean we can't go for that optimization.

In that case, it we are not handling this, the question remains if it is OK to delete a Section that is still used by permissions subsystem. From my POV that can be acceptable, as long as there is no Content involved. But, we should make sure that permissions are tolerant of inconsistencies like that. And it seems to me that should already be the case.

FWIW, Legacy Stack does not allow deleting a Section that is used by permissions, but OTOH it does allow deleting a ContentType that is used there.

Any thoughts? Ping @andrerom @bdunogier @dpobel

**Comment by glye** [comment]
Fair point - just noting that if we drop the policy check, we have to ensure role view/edit doesn't barf exceptions when it comes across limitations on non-existing sections. (Currently I don't think it will, as section limitations are numbers only at this point.)

**Comment by bdunogier** [comment]
What about moving the API level a bit: `canRemove( Section $section ): bool` ?

**Comment by andrerom** [comment]
I think I will have to side with Petar here, it is not very user friendly to have to clean up whole system when deleting meta data like this.

So in the future we should rather aim to give a UI to replace limitation and section assignment on deletion, but for now: how did legacy handle this?

**Comment by glye** [comment]
Legacy checks objects, policy limitations, and assignment limitations to determine if the section can be removed:
https://github.com/ezsystems/ezpublish-legacy/blob/master/kernel/classes/ezsection.php#L179

**Comment by andrerom** [comment]
:/ :p 

Well as I don't fully see why we would want to change storage to use json here:

+1


However do we need to expose this in API?

**Comment by glye** [comment]
It should be possible to limit it to the SPI.

**Comment by andrerom** [comment]
> It should be possible to limit it to the SPI.

Ok, but let's leave it like it is, could be useful for UI and if we regret we can deprecate in the future.

**Comment by andrerom** [comment]
so still +1

review ping @lolautruche

**Comment by lolautruche** [comment]
> What about moving the API level a bit: `SectionService::canRemove( Section $section ): bool` ?

I second @bdunogier's proposal, but then we should do the same for ContentType and possibly in other places.
The code looks OK, but this is API, so it is meant to be supported for a long time with BC. In that regard I'd better go with `SectionService::canRemove()`.

**Comment by andrerom** [comment]
> What about moving the API level a bit: SectionService::canRemove( Section $section ): bool ?

I'm -1 on that, as it implies permissions, if so lets keep this at SPI level only and skip doing APi changes here.

**Comment by glye** [comment]
@andrerom @lolautruche @bdunogier I think I have to change API since the SPI handler is protected in the the section service, and I need to access it from the section controller. Or can I make a `SectionService::_canRemoveHelper()` which is public but not part of the API, documented as subject to change?

**Comment by andrerom** [comment]
Well,

I'm +1 on `countPoliciesUsing` for API, so if @bdunogier and @lolautruche can live with that then you have the exposed api you need. `can*` implies permissions, which is a separate service, while this is business logic.

**Comment by andrerom** [comment]
Clarified a few things with @bdunogier to get the gist of what he and @lolautruche had in mind, so in total what we discussed implies the following:
- Remove `countPoliciesUsing()` from API
- Deprecate `countAssignedContents()` from API
- Expose something like `sectionUsed() : bool` or `sectionDeletable() : bool` *(if someone wants to discuss naming, go ahead, but don't involve me in it, I have a slight preferance towards the first form as it is more specific on it's meaning)* Make sure to document what it checks, and that it is not checking user permissions.
- In this new function, check countPoliciesUsing and countAssignedContents spi calls.

**Comment by glye** [comment]
Thanks guys, will progress it.

**Comment by bdunogier** [comment]
Please ignore then @andrerom :-)

I have a feeling that there is an overall concept we could clarify at some point: the fact that something can be deleted or not, in terms of business rules, is at least as important as permissions.

There is overall, I think, a need to cover this "deletable" thing. It potentially applies to all domain objects, and the Repository is, in most case, the only one able to inform about that. But I'm thinking that a "deletable" boolean might not be sufficient to satisfy all of our needs here.

One example: what if, if a user wants to delete a section, we offered the option to remap the policies that uses that section to another section, so that it _can_ be deleted ? Is it the application's responsibility to check the existing data, or the repository's to tell about that ? If it's the repository's, a boolean would not be sufficient there, as a more complicated message would have to be returned. Maybe a value object of some sorts ?

I'd still go for `isSectionDeletable()`. It is, by name, a yes/no question. We're not asking what needs to be done to make it deletable. If you prefer `isSectionUsed()`, I'm fine with that as well. But in both cases, it should also check if there is content using the section, should it not ? Is this already covered in another area ?

**Comment by glye** [comment]
@bdunogier The existing `countAssignedContents()` returns the number of content using the section. The proposed `(is)sectionUsed()` would check that, in addition to the usage in policies.

I don't think the example about remapping policies is in any big demand. Deleting sections is rarely done, and editing policies manually is not hard. The expection would be if you have tons of programmatically generated policies - presumably if people do that, they will also be able to programmatically remap policies in such cases.

/me don't want to make things more complex than necessary, though I see your point about a unified approach to `deletable` objects. (Of course, adding an `(is)sectionUsed()` now, which only checks usage, doesn't stop us from adding a general `isDeletable()` later, which also checks permissions.)

**Comment by crevillo** [comment]
my two cents...if the goal is to disable the delete button if the section can not be removed (say deleted) i'd go for ```isDeletable```. Thinking a section can be used but still be deletable, so no need to disable button in this case... 
or are we aiming to prevent the user to delete section whenever the section is used?

**Comment by andrerom** [comment]
> or are we aiming to prevent the user to delete section whenever the section is used?

No, what we want to do as a followup later is to detect that Section is in use, to provide UI to let user deal with migrating the data before deleting. Hence why "used" as opposed to "deletable" *(which implies permissions check)* makes most sense ;) /cc @bdunogier @glye

**Comment by crevillo** [comment]
Ok. Then i think the two methods can be added...

**Comment by andrerom** [comment]
Which two methods? The original once to be able to do such ui logic, maybe yes.

**Comment by crevillo** [comment]
```isSectionUsed``` to check is section is used in some way and ```isSectionDeletable``` to check if section can be deleted even if it's used...

**Comment by glye** [comment]
Added count of usage in role assignment limitations, so all 3 usage cases are covered.
Implemented in `isSectionUsed()` as discussed. An `isSectionDeletable()` would be a later follow-up IMHO, if we go for that.

File: eZ/Publish/Core/Persistence/Legacy/Content/Section/Gateway/ExceptionConversion.php [hunk 5]
```diff
@@ -145,6 +145,42 @@ public function countContentObjectsInSection($id)
     }
 
     /**
+     * Counts the number of role policies using section with $id in their limitations.
+     *
+     * @param int $id
+     *
+     * @return int
```
Comment (crevillo) [inline eZ/Publish/Core/Persistence/Legacy/Content/Section/Gateway/ExceptionConversion.php]: should we add documentation about the exceptions thrown?

File: eZ/Publish/SPI/Persistence/Content/Section/Handler.php [hunk 6]
```diff
@@ -94,4 +94,22 @@ public function assign($sectionId, $contentId);
...
+    public function policiesCount($sectionId);
+
+    /**
+     * Counts the number of role assignments using section with $sectionId in their limitations.
+     *
+     * @param int $sectionId
+     *
+     * @return int
+     */
+    public function roleAssignmentsUsingSectionCount($sectionId);
```
Comment (crevillo) [inline eZ/Publish/SPI/Persistence/Content/Section/Handler.php]: UsingSectionCount? sounds like we're getting role Assigments using a "Section Count" as if count was some attr of the section like the id... don't you think? i'd go for ```countRoleAssignmentsUsingSection``` maybe...

Comment (glye) [inline eZ/Publish/SPI/Persistence/Content/Section/Handler.php]: It's done to be consistent with `assignmentsCount` and `policiesCount`. I have no strong opinions on the matter :)

File: eZ/Publish/Core/Persistence/Legacy/Content/Section/Gateway/ExceptionConversion.php (hunk 5 above)
Comment (glye) [inline eZ/Publish/Core/Persistence/Legacy/Content/Section/Gateway/ExceptionConversion.php]: We don't do that in any of the other `ExceptionConversion` classes either afaict, so I'd deal with that in a separate issue if we're doing it. Don't know why it has been skipped.

File: eZ/Publish/API/Repository/SectionService.php [hunk 7]
```diff
@@ -84,10 +84,23 @@ public function loadSectionByIdentifier($sectionIdentifier);
...
     public function countAssignedContents(Section $section);
 
     /**
+     * Returns true if the given section is assigned to contents, or used in role policies, or in role assignments.
+     *
+     * This does not check user permissions.
+     *
+     * @param \eZ\Publish\API\Repository\Values\Content\Section $section
+     *
+     * @return bool
```
Comment (andrerom) [inline eZ/Publish/API/Repository/SectionService.php]: Add "@since 6.0"

File: eZ/Publish/SPI/Persistence/Content/Section/Handler.php (hunk 6 above)
Comment (andrerom) [inline eZ/Publish/SPI/Persistence/Content/Section/Handler.php]: The original convention is countContentObjectsUsingSection, can maybe reflect that form on the new once also.

**Comment by andrerom** [comment]
+1 besides 2 comments

**Comment by glye** [comment]
Changed to `countContentObjectsUsingSection()` all over
Added `@since 6.0`

**Comment by bdunogier** [comment]
+1 as far as I am concerned. I think we got ourselves something clean.

**Comment by bdunogier** [comment]
> "deletable" (which implies permissions check)
It doesn't to me. We already have "can" that is reserved to permissions. We can't _also_ give permissions the "-able" prefix :-)

It's not important now, but we will have to draw a semantic line somewhere.

**Comment by glye** [comment]
I'm still working on the tests, and the dependency https://github.com/ezsystems/PlatformUIBundle/pull/415 isn't approved yet.

**Comment by glye** [comment]
Added tests.

**Pull request merged**